plugins/catalog-backend/src/ingestion/processors/BitbucketDiscoveryProcessor.ts

```diff
@@ -0,0 +1,146 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Logger } from 'winston';
+import { Config } from '@backstage/config';
+
+import { ScmIntegrations } from '@backstage/integration';
+import { LocationSpec } from '@backstage/catalog-model';
+import { BitbucketClient, pageIterator } from './bitbucket';
+import { CatalogProcessor, CatalogProcessorEmit } from './types';
+import { results } from './index';
+
+export class BitbucketDiscoveryProcessor implements CatalogProcessor {
+  private readonly integrations: ScmIntegrations;
+  private readonly logger: Logger;
+
+  static fromConfig(config: Config, options: { logger: Logger }) {
+    const integrations = ScmIntegrations.fromConfig(config);
+
+    return new BitbucketDiscoveryProcessor({
+      ...options,
+      integrations,
+    });
+  }
+
+  constructor(options: { integrations: ScmIntegrations; logger: Logger }) {
+    this.integrations = options.integrations;
+    this.logger = options.logger;
+  }
+
+  async readLocation(
+    location: LocationSpec,
+    _optional: boolean,
+    emit: CatalogProcessorEmit,
+  ): Promise<boolean> {
+    if (location.type !== 'bitbucket-discovery') {
+      return false;
+    }
+
+    const bitbucketConfig = this.integrations.bitbucket.byUrl(location.target)
+      ?.config;
+    if (!bitbucketConfig) {
+      throw new Error(
+        `There is no Bitbucket integration that matches ${location.target}. Please add a configuration entry for it under integrations.bitbucket`,
+      );
+    } else if (bitbucketConfig.host === 'bitbucket.org') {
+      throw new Error(
+        `Component discovery for Bitbucket Cloud is not yet supported`,
+      );
+    }
+
+    const client = new BitbucketClient({
+      config: bitbucketConfig,
+    });
+    const startTimestamp = Date.now();
+    this.logger.info(`Reading Bitbucket repositories from ${location.target}`);
+
+    const repositories = await readBitbucketOrg(client, location.target);
+
+    for (const repository of repositories) {
+      emit(
+        results.location(
+          repository,
+          // Not all locations may actually exist, since the user defined them as a wildcard pattern.
+          // Thus, we emit them as optional and let the downstream processor find them while not outputting
+          // an error if it couldn't.
+          true,
+        ),
+      );
+    }
+
+    const duration = ((Date.now() - startTimestamp) / 1000).toFixed(1);
+    this.logger.info(
+      `Read ${repositories.length} Bitbucket repositories in ${duration} seconds`,
```

Comment:
we recently reshuffled slightly in the github one, so that it shows not just how many were read, but how many that actually matched too, probably good to do here too https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts#L85

Reply:
The readBitbucketOrg function only returns matched repositories. However, I have updated the returned object to also include the total number of scanned repositories, which is then used for logging.